File: ios/SIOPKit/Sources/SIOPKit/JWS.swift
import Foundation

/// Minimal JWS (RFC 7515) compact serialization with RS256, the mandatory
/// signing algorithm for Self-Issued OPs (OpenID Connect Core 1.0 Section 7.1).
enum JWS {
    static func signRS256(payload: [String: Any], key: SIOPKeyProvider) throws -> String {
        let header = try encodeSegment(["alg": "RS256", "typ": "JWT"])
        let body = try encodeSegment(payload)
        let signingInput = "\(header).\(body)"
        let signature = try key.sign(Data(signingInput.utf8))
        return "\(signingInput).\(Base64URL.encode(signature))"
    }

    static func decode(
        _ jwt: String
    ) throws -> (header: [String: Any], payload: [String: Any], signature: Data, signingInput: Data) {
        let parts = jwt.components(separatedBy: ".")
        guard parts.count == 3,
              let headerData = Base64URL.decode(parts[0]),
              let payloadData = Base64URL.decode(parts[1]),
              let signature = Base64URL.decode(parts[2]),
              let header = (try? JSONSerialization.jsonObject(with: headerData)) as? [String: Any],
              let payload = (try? JSONSerialization.jsonObject(with: payloadData)) as? [String: Any]
        else {
            throw SIOPError.invalidToken("malformed JWT")
        }
        return (header, payload, signature, Data("\(parts[0]).\(parts[1])".utf8))
    }

    private static func encodeSegment(_ object: [String: Any]) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: object, options: [.sortedKeys])
        return Base64URL.encode(data)
    }
}

File: android/siopkit/src/main/kotlin/jp/co/pendako/siop/Jws.kt
package jp.co.pendako.siop

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Minimal JWS (RFC 7515) compact serialization with RS256, the signing
 * algorithm a Self-Issued OP must support (OpenID Connect Core 1.0 Section 7.1).
 */
internal object Jws {
    private val json = Json { encodeDefaults = true }

    fun signRs256(payload: JsonObject, key: SiopKeyProvider): String {
        val header = buildJsonObject {
            put("alg", "RS256")
            put("typ", "JWT")
        }
        val signingInput = "${encode(header)}.${encode(payload)}"
        return "$signingInput.${Base64Url.encode(key.sign(signingInput.toByteArray()))}"
    }

    fun decode(jwt: String): DecodedJws {
        val parts = jwt.split(".")
        if (parts.size != 3) throw SiopError.InvalidToken("malformed JWT")
        return try {
            DecodedJws(
                header = json.parseToJsonElement(String(Base64Url.decode(parts[0]))) as JsonObject,
                payload = json.parseToJsonElement(String(Base64Url.decode(parts[1]))) as JsonObject,
                signature = Base64Url.decode(parts[2]),
                signingInput = "${parts[0]}.${parts[1]}".toByteArray(),
            )
        } catch (cause: Exception) {
            throw SiopError.InvalidToken("malformed JWT")
        }
    }

    private fun encode(value: JsonObject): String =
        Base64Url.encode(json.encodeToString(JsonObject.serializer(), value).toByteArray())
}

internal class DecodedJws(
    val header: JsonObject,
    val payload: JsonObject,
    val signature: ByteArray,
    val signingInput: ByteArray,
)
